benchmarks/src/main/java/io/grpc/benchmarks/qps/OpenLoopClient.java

```diff
@@ -0,0 +1,237 @@
+    final Histogram histogram = new AtomicHistogram(60000000, 3);
+    final TestServiceStub stub;
+    final SimpleRequest request;
+    final Random rnd;
+    final int targetQps;
+    final long numRpcs;
+
+    LoadGenerationWorker(Channel channel, SimpleRequest request, int targetQps, int duration) {
+      stub = TestServiceGrpc.newStub(checkNotNull(channel, "channel"));
+      this.request = checkNotNull(request, "request");
+      this.targetQps = targetQps;
+      numRpcs = targetQps * duration;
+      rnd = new Random();
+    }
+
+    /**
+     * Discuss waiting strategy between calls. Sleeping seems to be very inaccurate
+     * (see below). On the other hand calling System.nanoTime() a lot (especially from
+     * different threads seems to impact its accuracy
+     * http://shipilev.net/blog/2014/nanotrusting-nanotime/
+     * On my system the overhead of LockSupport.park(long) seems to average at ~55 micros.
+     * // Try to sleep for 1 nanosecond and measure how long it actually takes.
+     * long start = System.nanoTime();
+     * int i = 0;
+     * while (i < 10000) {
+     *   LockSupport.parkNanos(1);
+     *   i++;
+     * }
+     * long end = System.nanoTime();
+     * System.out.println((end - start) / 10000);
+     */
+    @Override
+    public Histogram call() throws Exception {
+      long now = System.nanoTime();
+      long nextRpc = now;
+      long i = 0;
+      while (i < numRpcs) {
+        now = System.nanoTime();
+        if (nextRpc - now <= 0) {
+          // TODO: Add option to print how far we have been off from the target delay in micros.
+          nextRpc += nextDelay(targetQps);
+          newRpc(stub);
+          i++;
+        }
+      }
+
+      waitForRpcsToComplete(1);
+
+      return histogram;
+    }
+
+    private void newRpc(TestServiceStub stub) {
+      stub.unaryCall(request, new StreamObserver<Qpstest.SimpleResponse>() {
+
+        private final long start = System.nanoTime();
+
+        @Override
+        public void onValue(Qpstest.SimpleResponse value) {
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          Status status = Status.fromThrowable(t);
+          System.err.println("Encountered an error in unaryCall. Status is " + status);
+          t.printStackTrace();
+        }
+
+        @Override
+        public void onCompleted() {
+          final long end = System.nanoTime();
+          histogram.recordValue((end - start) / 1000);
+        }
+      });
+    }
+
+    private void waitForRpcsToComplete(int duration) throws InterruptedException {
+      long now = System.nanoTime();
+      long end = now + duration * 1000 * 1000 * 1000;
+      while (histogram.getTotalCount() < numRpcs && now < end) {
+        now = System.nanoTime();
+      }
+    }
+
+    private long nextDelay(int lambda) {
+      //
+      double seconds = -Math.log(rnd.nextDouble()) / lambda;
+      double nanos = seconds * 1000 * 1000 * 1000;
```

Comment:
@nmittler `seconds` may often be well below 1, so I can't just cast here. The `Math.round` will not make a diff, as it's only +/- 1 ns and I am very happy if we can be micros accurate already :-).